### ISSUE TITLE
Allow add-ons to download other add-ons as dependencies

### ISSUE THREAD

**Issue by Morgy93** [ddev/ddev#5337]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem?

In addition to https://github.com/ddev/ddev/issues/4717 I'd like the ability to download dependencies automatically.

If you have multiple dependencies (e.g. in a metapackage) it is tedious that you must manually install and update the dependencies.

### Describe your solution

Therefore, I'd like some global setting "allow automatic add-on dependency management" and corresponding features.

### Describe alternatives

Since DDEV is heavily focused on PHP it might be an option to use an established package manager for it like https://getcomposer.org/

**Comment by GuySartorelli** [comment]
DDEV add-ons don't really have "versions" per se - maybe they will and the dependencies will include version constraints in the future, but for now it seems simple enough to have a `--include-dependencies` flag as part of `ddev get`, which if passed in would result in recursively including missing dependencies instead of outputting the message about missing dependencies.

@rfay Does that sound like a good approach?
@Morgy93 would that be suitable for your needs?

If both of you say yes I'd be happy to work on this.

**Comment by Morgy93** [comment]
> [...] it seems simple enough to have a `--include-dependencies` flag as part of `ddev get`, which if passed in would result in recursively including missing dependencies instead of outputting the message about missing dependencies. [...]
>
>@Morgy93 would that be suitable for your needs?

Sound great to me 👍

**Comment by rfay** [comment]
> for now it seems simple enough to have a --include-dependencies

I think people will just assume that it should use `--include-dependencies` and skip adding that flag. The one complexity there is if they already have the dependency it will be upgraded, but that's probably not going to be a big problem either.

**Comment by GuySartorelli** [comment]
> I think people will just assume that it should use `--include-dependencies` and skip adding that flag.

I can make it the default if you like (so add a `--no-dependencies` or similarly named flag to out _out_ instead).
Alternatively I'd update the message that currently displays when dependencies are missing to say "run the command again with the `--include-dependencies` flag or manually add those dependencies first" or something along those lines.

> The one complexity there is if they already have the dependency it will be upgraded, but that's probably not going to be a big problem either.

I don't think that's necessarily true? Certainly it could be made to work that way, but my assumption is it would check for missing dependencies (which it already does) and just recursively install anything that's missing. If the dependency is already installed I wouldn't expect it to be touched.

**Comment by rfay** [comment]
It would be OK that way, but I do think people will just want it to work in the simplest way. Maybe output what's happening and say how to revert it (with versions)

**Comment by GuySartorelli** [comment]
> It would be OK that way

Is this about the `--include-dependencies` flag, the `--no-dependencies` flag, or installing missing dependencies rather than updating them?

**Comment by rfay** [comment]
My preference is to *not* do a new flag, just download dependencies.

If a dependency is being upgraded, output information about the change.

**Comment by GuySartorelli** [comment]
> If a dependency is being upgraded, output information about the change.

Just to be clear, you do want dependencies to be upgraded if there's an upgrade to be made, rather than just recursively adding missing dependencies?

Note that I think that'll actually be a little more difficult to implement than just adding the missing ones.

**Comment by rfay** [comment]
Yes, just do a "get" on needed dependencies. The naive get that does the default latest release.

What do you think @stasadev ?

**Comment by stasadev** [comment]
I don't mind, let's encourage people to get the latest.

**Comment by GuySartorelli** [comment]
I was about to start working on this when I realised that one way people can use `ddev get` is to install from a directory.

Is is possible to be in a scenario where you have installed a local version of a dependency already, and then you run `ddev get something-else` and `something-else` has a dependency with the same name as the thing you installed from a local directory, and that results in replacing the one you intentionally installed from a local copy with the most recent internet copy?

i.e.
1. install "dependencyA" from local directory
2. install "dependecyB" which lists "dependencyA" as a dependency
  - Results in re-installing "dependencyA"  from the internet, replacing the copy you had installed earlier

It might seem like an unlikely edge case but it's entirely possible if someone maintains both dependencies and is making changes to one of them, and validating that those changes don't affect the other one.

With the current suggested implementation, we would just do as written above (install the latest version of both dependencies) and the output would reflect that - but I just want to make sure that's the behaviour you want before I implement it.

**Comment by tyler36** [comment]
I know the `tyler36/ddev-xhgui` addon wants to have a dependency of `ddev-mongo` so keen to figure something out.

Perhaps we could check if the dependency is already installed. 
If not, grab the latest.
If so, display message suggesting the developer manually updates. They may be a valid reason why they are on an older/custom version.

I don't think many people are installing via path though. I do mostly when I create or update an addon. I think the path-install is also used if the addon is in a private repo (?).

**Comment by rfay** [comment]
A complexity to be resolved here is that `dependencies` currently is a reference to the `name` of the add-on, for example `xhgui` in https://github.com/ddev/ddev-xhgui/blob/main/install.yaml#L1

However, the `name` is not enough for us to download the item, it has to be either a github reference (like `ddev/ddev-xhgui`), a tarball URL (like https://github.com/ddev/ddev-xhgui/tarball/main), or a filepath.

**Comment by rfay** [comment]
This could be a good candidate for a live demonstration screencast/training PR using claude.
